### PR TITLE
New Script: Zammad

### LINKED ISSUE #642: Zammad Ticketsystem
https://github.com/community-scripts/ProxmoxVE/discussions/108

Application Name
Zammad

Website
https://docs.zammad.org/en/latest/install/package.html

Description
Zammad is a web based open source helpdesk/customer support system with many features to manage customer communication via several channels like telephone, facebook, twitter, chat and emails. It is distributed under version 3 of the GNU AFFERO General Public License (GNU AGPLv3).

### PR DESCRIPTION
**Pull request by michelroegl-brunner**
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description

Provide a summary of the changes made and/or reference the issue being addressed.

Fixes #642 

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [X] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [X] Testing performed (I have tested my changes, ensuring everything works as expected)
- [ ] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)
Provide any additional context or screenshots about the feature or fix here.

## Related Pull Requests / Discussions

If there are other pull requests or discussions related to this change, please link them here:
- Related PR #

### CODE REVIEW

File: ct/zammad.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+source <(curl -s https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/zammad/misc/build.func)
+
```
Comment (MickLesk) [inline ct/zammad.sh:3]: remove blank line

File: ct/zammad.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#Copyright (c) 2021-2024 community-scripts ORG
```
Comment (MickLesk) [inline ct/zammad.sh:4]: Insert space

File: ct/zammad.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+# Author: Michel Roegl-Brunner (michelroegl-brunner)
+# License: MIT
+# https://github.com/community-scripts/ProxmoxVE/raw/main/LICENSE
+function header_info {
```
Comment (MickLesk) [inline ct/zammad.sh:8]: Insert empty line before it

File: ct/zammad.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+if [[ ! -d /opt/zamad ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
+msg_info "Updating ${APP} LXC"
+systemctl stop zammad &>/dev/null
+apt-get update &>/dev/null
+apt-mark hold zammad &>/dev/null
+apt-get -y upgrade &>/dev/null
+apt-mark unhold zammad &>/dev/null
+apt-get -y upgrade &>/dev/null
+systemctl start zammad &>/dev/null
+msg_ok "Updated ${APP} LXC"
```
Comment (MickLesk) [inline ct/zammad.sh:67]: is zammad really in /opt? 

```bash
msg_info "Stopping ${APP}"
systemctl stop _APP_
msg_ok "Stopped ${APP}"

msg_info "Updating LXC packages"
apt-get update &>/dev/null
apt-get -y upgrade &>/dev/null
msg_ok "Updated LXC packages"

msg_info "Updating ${APP}"
cd /opt
DO SOMETHING UPDATE
DO SOMETHING UPDATE
msg_ok "Updated ${APP}"

msg_info "Starting ${APP}"
systemctl start _APP_
msg_ok "Started ${APP}"

msg_ok "Updated Successfully"
```

File: install/zammad-install.sh [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+msg_info "Installing Dependencies"
+$STD apt-get install -y \
+  curl \
+  git \
+  sudo \
+  gpg \
+  wget \
+  nginx
+
```
Comment (MickLesk) [inline install/zammad-install.sh:24]: msg_ok "Installed Dependencies"

msg_info "Setting up ElasticSearch"

File: install/zammad-install.sh [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+echo "deb [signed-by=/etc/apt/trusted.gpg.d/elasticsearch.gpg] https://artifacts.elastic.co/packages/7.x/apt stable main"| tee -a /etc/apt/sources.list.d/elastic-7.x.list > /dev/null
+curl -fsSL https://artifacts.elastic.co/GPG-KEY-elasticsearch | gpg --dearmor | tee /etc/apt/trusted.gpg.d/elasticsearch.gpg> /dev/null
+$STD apt-get update
+$STD apt-get -y install elasticsearch
+echo "-Xms2g" >> /etc/elasticsearch/jvm.options
+echo "-Xmx2g" >> /etc/elasticsearch/jvm.options
+$STD /usr/share/elasticsearch/bin/elasticsearch-plugin install ingest-attachment -b
+$STD systemctl restart elasticsearch
+msg_ok "Installed Dependencies"
```
Comment (MickLesk) [inline install/zammad-install.sh:33]: msg_ok "Setup ElasticSearch"

File: install/zammad-install.sh [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+msg_info "Install Zammad"
```
Comment (MickLesk) [inline install/zammad-install.sh:35]: msg_info "Installing Zammad"

File: install/zammad-install.sh [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+curl -fsSL https://dl.packager.io/srv/zammad/zammad/key | gpg --dearmor | sudo tee /etc/apt/keyrings/pkgr-zammad.gpg> /dev/null
+echo "deb [signed-by=/etc/apt/keyrings/pkgr-zammad.gpg] https://dl.packager.io/srv/deb/zammad/zammad/stable/debian 12 main"| sudo tee /etc/apt/sources.list.d/zammad.list > /dev/null
+$STD apt-get update
+$STD apt-get -y install zammad
+
+msg_ok "Installed Zammad"
+
+msg_info "Configuring Servcie"
```
Comment (MickLesk) [inline install/zammad-install.sh:44]: remove this 3 lines

File: install/zammad-install.sh [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+$STD zammad run rails r "Setting.set('es_url', 'http://localhost:9200')"
+$STD zammad run rake zammad:searchindex:rebuild
+
```
Comment (MickLesk) [inline install/zammad-install.sh:47]: msg_ok "Installed Zammad"

msg_info "Set up web services"

File: install/zammad-install.sh [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+cp /opt/zammad/contrib/nginx/zammad.conf /etc/nginx/sites-available/zammad.conf
+IPADDRESS=$(hostname -I | awk '{print $1}')
+sed -i "s/server_name localhost;/server_name $IPADDRESS;/g" /etc/nginx/sites-available/zammad.conf
+$STD systemctl reload nginx
+
```
Comment (MickLesk) [inline install/zammad-install.sh:52]: msg_ok "Created Services"

File: install/zammad-install.sh [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+msg_info "Installing Dependencies"
+$STD apt-get install -y \
+  curl \
+  git \
+  sudo \
+  gpg \
+  wget \
+  nginx
+
+echo "deb [signed-by=/etc/apt/trusted.gpg.d/elasticsearch.gpg] https://artifacts.elastic.co/packages/7.x/apt stable main"| tee -a /etc/apt/sources.list.d/elastic-7.x.list > /dev/null
```
Comment (MickLesk) [inline install/zammad-install.sh:25]: no /dev/null

File: install/zammad-install.sh [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+curl -fsSL https://artifacts.elastic.co/GPG-KEY-elasticsearch | gpg --dearmor | tee /etc/apt/trusted.gpg.d/elasticsearch.gpg> /dev/null
```
Comment (MickLesk) [inline install/zammad-install.sh:26]: no /dev/null

curl -fsSL https://artifacts.elastic.co/GPG-KEY-elasticsearch | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg

File: ct/zammad.sh (hunk 4 above)
Comment (MickLesk) [reply ct/zammad.sh:67]: Then can you Split Like other Scripts?

MSG Stop Service

MSG Update App

MSG start Service ?

File: install/zammad-install.sh [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+msg_info "Setting up Elasticsearch"
+echo "deb [signed-by=/etc/apt/trusted.gpg.d/elasticsearch.gpg] https://artifacts.elastic.co/packages/7.x/apt stable main"| tee -a /etc/apt/sources.list.d/elastic-7.x.list 
+curl -fsSL https://artifacts.elastic.co/GPG-KEY-elasticsearch | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
+$STD apt-get update
+$STD apt-get -y install elasticsearch
+echo "-Xms2g" >> /etc/elasticsearch/jvm.options
+echo "-Xmx2g" >> /etc/elasticsearch/jvm.options
+$STD /usr/share/elasticsearch/bin/elasticsearch-plugin install ingest-attachment -b
+$STD systemctl restart elasticsearch
```
Comment (MickLesk) [inline install/zammad-install.sh:34]: -q without Std?

File: install/zammad-install.sh [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+msg_info "Installing Zammad"
+curl -fsSL https://dl.packager.io/srv/zammad/zammad/key | gpg --dearmor | sudo tee /etc/apt/keyrings/pkgr-zammad.gpg> /dev/null
+echo "deb [signed-by=/etc/apt/keyrings/pkgr-zammad.gpg] https://dl.packager.io/srv/deb/zammad/zammad/stable/debian 12 main"| sudo tee /etc/apt/sources.list.d/zammad.list > /dev/null
+$STD apt-get update
+$STD apt-get -y install zammad
+$STD zammad run rails r "Setting.set('es_url', 'http://localhost:9200')"
+$STD zammad run rake zammad:searchindex:rebuild
+msg_ok "Installed Zammad"
+
+msg_info "Set up web services"
```
Comment (MickLesk) [inline install/zammad-install.sh:46]: Setup Services

File: install/zammad-install.sh (hunk 13 above)
Comment (michelroegl-brunner) [reply install/zammad-install.sh:34]: You mean the plugin install? Sadly no

Comment (MickLesk) [reply install/zammad-install.sh:34]: I mean the systemctl

**Comment by michelroegl-brunner** [comment]
Updated the PR to account for the changes. You can test it on my dev branch: bash -c "$(wget -qLO - https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/DEV/ct/zammad.sh)"

**Comment by MickLesk** [comment]
> Updated the PR to account for the changes. You can test it on my dev branch: bash -c "$(wget -qLO - https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/DEV/ct/zammad.sh)"

nope

image

**Comment by michelroegl-brunner** [comment]
> > Updated the PR to account for the changes. You can test it on my dev branch: bash -c "$(wget -qLO - https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/DEV/ct/zammad.sh)"
> 
> nope
> 
> image

Can you test again if this was an isolated error? I´ve run the script on 3 hosts now without a problem.
zammad

**Comment by litecactus** [comment]
I'm new to these community scripts but have a successful build on my server.
What is the default container console username and password please?
IP address given after script runs shows a `This site can’t be reached
192.168.1.21 refused to connect.`

**Comment by MickLesk** [comment]
> I'm new to these community scripts but have a successful build on my server. What is the default container console username and password please? IP address given after script runs shows a `This site can’t be reached 192.168.1.21 refused to connect.` though I would expect to be able to connect

if you get an user passwort question, your lxc is defect!

**Comment by MickLesk** [comment]
@michelroegl-brunner Switch CT to new Default

**Review by bramsuurdje — DISMISSED** [review]

**Comment by michelroegl-brunner** [comment]
@MickLesk Done!

**Review by MickLesk — DISMISSED** [review]

**Review by MickLesk — DISMISSED** [review]

**Comment by litecactus** [comment]
Works for me, thank you

**Review by MickLesk — APPROVED** [review]

**Comment by second2falcon** [comment]
@MickLesk sorry for the late response, but it worked for me fine too.

**Comment by chrisjenx** [comment]
Just ran this today, failed to setup correctly, the update/running the script again will also say "no installation found"

**Comment by michelroegl-brunner** [comment]
Please open a issue and fill out the form and dont answer on closed Pull Requests. Thank you!

**Comment by chrisjenx** [comment]
Nvm, the issue is with Nginx, should probably create an option in the script to just expose directly so we can use our own reverse proxy

**Pull request merged**